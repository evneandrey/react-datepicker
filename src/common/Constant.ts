export const DatePickerDefaults = {
  dateFormat: 'YYYY-MM-DD',
  dateTimeFormat: 'YYYY-MM-DD HH:mm A',
  timeFormat: 'HH:mm A',
  locale: 'en',
};
