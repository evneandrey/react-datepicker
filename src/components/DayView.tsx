import * as classNames from 'classnames';
import * as dayjs from 'dayjs';
import * as React from 'react';
import { DatePickerDefaults } from '../common/Constant';
import TableCell from './TableCell';
import TableMatrixView from './TableMatrixView';
import { ifExistCall } from '../utils/FunctionUtil';
import { getWeekDays } from '../utils/LocaleUtil';

import { getDayMatrix, isDayEqual, isDayRange } from '../utils/DateUtil';
import { IDatePicker } from '../common/@types';

export interface Props {
  /** Selected days to show in calendar */
  selected?: dayjs.Dayjs[];
  /** Start day to show in calendar */
  startDay?: dayjs.Dayjs;
  /** End day to show in calendar */
  endDay?: dayjs.Dayjs;
  /** Calendar day click Event */
  onClick?: (date: string) => void;
  /** Calendar day Mouseover Event */
  onMouseOver?: (date: dayjs.Dayjs) => void;
  /** Custom day class to show in day */
  customDayClass?: (date: dayjs.Dayjs) => string | string[];
  /** Custom day text to show in day */
  customDayText?: (date: dayjs.Dayjs) => string;
  /** Calendar day disable */
  disableDay?: (date: dayjs.Dayjs) => void;
}

interface PrivateProps {
  current: dayjs.Dayjs;
  locale: IDatePicker.Locale;
}

class DayView extends React.Component<Props & PrivateProps> {
  public static defaultProps = {
    locale: DatePickerDefaults.locale,
  };

  public getDayClass = (date: string): string => {
    const { current, customDayClass, startDay, endDay, selected, disableDay } = this.props;
    const currentDate = dayjs(current).date(parseInt(date, 10));

    let classArr: string[] = [];

    if (!date.trim()) {
      return '';
    }

    if (customDayClass !== undefined) {
      const customClass = customDayClass(currentDate);
      classArr = classArr.concat(typeof customClass === 'string' ? [customClass] : customClass);
    }

    const dayClass = classNames(
      'calendar__day',
      `calendar__day--${dayjs(currentDate).day()}`,
      classArr,
      {
        'calendar__day--end': isDayEqual(currentDate, endDay),
        'calendar__day--range': isDayRange(currentDate, startDay, endDay),
        'calendar__day--selected': this.isIncludeDay(date, selected),
        'calendar__day--disabled': disableDay ? disableDay(currentDate) : false,
        'calendar__day--start': isDayEqual(currentDate, startDay),
        'calendar__day--today': isDayEqual(currentDate, dayjs()),
      }
    );

    return dayClass;
  };

  public getCustomText = (date: string): string => {
    const { current, customDayText } = this.props;
    const currentDate = dayjs(current).date(parseInt(date, 10));

    if (!date.trim()) {
      return '';
    }
    if (!customDayText) {
      return '';
    }

    return customDayText(currentDate);
  };

  public isIncludeDay = (date: string, dates?: dayjs.Dayjs[]): boolean => {
    const { current } = this.props;
    if (dates === undefined) {
      return false;
    }
    return dates.some(v => isDayEqual(dayjs(current).date(parseInt(date, 10)), v));
  };

  public handleClick = (date: string) => {
    const { current, disableDay } = this.props;
    const currentDate = dayjs(current).date(parseInt(date, 10));
    if (!(disableDay && disableDay(currentDate))) {
      ifExistCall(this.props.onClick, date);
    }
  };

  public handleMouseOver = (date: string) => {
    const { onMouseOver, current } = this.props;
    ifExistCall(onMouseOver, dayjs(current).date(parseInt(date, 10)));
  };

  public render() {
    const { current, locale } = this.props;

    const dayMatrix = getDayMatrix(dayjs(current).year(), dayjs(current).month());
    const weekdays = getWeekDays(locale);

    return (
      <TableMatrixView
        headers={weekdays}
        matrix={dayMatrix}
        cell={(date, key) => (
          <TableCell
            className={this.getDayClass(date)}
            subText={this.getCustomText(date)}
            onClick={this.handleClick}
            onMouseOver={this.handleMouseOver}
            text={date}
            key={key}
          />
        )}
      />
    );
  }
}

export default DayView;
