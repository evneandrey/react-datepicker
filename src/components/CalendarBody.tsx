import * as React from 'react';
import * as dayjs from 'dayjs';
import { IDatePicker } from '../common/@types';
import { DatePickerDefaults } from '../common/Constant';
import { getMonthMatrix, getYearMatrix } from '../utils/DateUtil';
import DayView, { Props as DayViewProps } from './DayView';
import TableCell from './TableCell';
import TableMatrixView from './TableMatrixView';

interface CalendarBodyProps {
  /** Calendar viewMode(Year, Month, Day) */
  viewMode: IDatePicker.ViewMode;
  /** Calendar current Date */
  current: dayjs.Dayjs;
  /** DayClick Event */
  onClick: (value: string) => void;
  /** Locale to use */
  locale: IDatePicker.Locale;
}
type Props = DayViewProps & CalendarBodyProps;

const YEAR_VIEW_CLASS = 'calendar__year';
const MONTH_VIEW_CLASS = 'calendar__month';

const buildMatrixView = (
  matrix: string[][],
  className: string,
  onClick: (key: number, value: string) => () => void
) => {
  return (
    <TableMatrixView
      matrix={matrix}
      cell={(value: string, key: number) => (
        <TableCell key={key} className={className} text={value} onClick={onClick(key, value)} />
      )}
    />
  );
};

class CalendarBody extends React.Component<Props> {
  public static defaultProps = {
    viewMode: IDatePicker.ViewMode.DAY,
    locale: DatePickerDefaults.locale,
  };

  public render() {
    const { current, onClick, locale } = this.props;
    const viewMap = {
      [IDatePicker.ViewMode.YEAR]: buildMatrixView(
        getYearMatrix(dayjs(current).year()),
        YEAR_VIEW_CLASS,
        (_, v) => () => onClick(v)
      ),
      [IDatePicker.ViewMode.MONTH]: buildMatrixView(
        getMonthMatrix(locale),
        MONTH_VIEW_CLASS,
        (k, _) => () => onClick(String(k))
      ),
      [IDatePicker.ViewMode.DAY]: <DayView {...this.props} />,
    };

    return <div className="calendar__body">{viewMap[this.props.viewMode]}</div>;
  }
}
export default CalendarBody;
