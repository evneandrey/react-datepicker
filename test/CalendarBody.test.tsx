import { mount, shallow, ReactWrapper, ShallowWrapper } from 'enzyme';
import * as sinon from 'sinon';
import * as dayjs from 'dayjs';
import * as React from 'react';
import DayView from '../src/components/DayView';
import CalendarBody from '../src/components/CalendarBody';
import { IDatePicker } from '../src/common/@types';

describe('<CalendarBody />', () => {
  const defaultProps = {
    current: dayjs(new Date(2018, 11, 1)),
    onClick: sinon.spy(),
  };
  let shallowComponent: ShallowWrapper<React.Component>;
  let mountComponent: ReactWrapper;

  it('should render correctly', () => {
    shallowComponent = shallow(<CalendarBody {...defaultProps} />);
    expect(shallowComponent).toBeTruthy();
    expect(shallowComponent).toMatchSnapshot();
  });

  describe('prop: viewMode', () => {
    it('should ViewMode.DAY correctly', () => {
      shallowComponent = shallow(
        <CalendarBody {...defaultProps} viewMode={IDatePicker.ViewMode.DAY} />
      );
      expect(shallowComponent.find(DayView)).toHaveLength(1);
    });

    it('should ViewMode.MONTH correctly', () => {
      mountComponent = mount(
        <CalendarBody {...defaultProps} viewMode={IDatePicker.ViewMode.MONTH} />
      );
      expect(mountComponent.find('td.calendar__month')).toHaveLength(12);
    });

    it('should ViewMode.YEAR correctly', () => {
      mountComponent = mount(
        <CalendarBody {...defaultProps} viewMode={IDatePicker.ViewMode.YEAR} />
      );
      expect(mountComponent.find('td.calendar__year')).toHaveLength(9);
    });
  });
});
