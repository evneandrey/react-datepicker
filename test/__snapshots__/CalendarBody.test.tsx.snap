// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarBody /> should render correctly 1`] = `
<div
  className="calendar__body"
>
  <DayView
    current={"2018-11-30T15:00:00.000Z"}
    locale="en"
    onClick={[Function]}
    viewMode={2}
  />
</div>
`;
